api/controllers/FlairController.js

```diff
@@ -5,122 +5,65 @@ var _ = require("lodash");
-          }
-        } else {
-          if (app.flair.indexOf('ribbon') > -1 && css_class && css_class.indexOf('ribbon') === -1) {
-            css_class = css_class + " " + app.flair;
-          } else if (app.flair.indexOf('ribbon') === -1 && css_class && css_class.indexOf('ribbon') > -1) {
-            css_class = app.flair + " " + css_class;
-          } else {
-            css_class = app.flair;
-          }
-        }
-        Reddit.setUserFlair(req.user.redToken, user.name, css_class, flair, app.sub).then(function () {
-          Event.create({
-            type: "flairTextChange",
-            user: req.user.name,
-            content: "Changed " + user.name + "'s flair to " + css_class
-          }).exec(function () {
-          });
-
-          console.log("/u/" + req.user.name + ": Changed " + user.name + "'s flair to " + css_class);
-          Reddit.sendPrivateMessage(
-            refreshToken,
-            'FlairHQ Notification',
-            'Your application for ' + formatted + ' flair on /r/' + app.sub + ' has been approved.',
-            user.name).then(undefined, function () {
-              console.log('Failed to send a confirmation PM to ' + user.name);
-            }
-          );
-          if (app.sub === 'pokemontrades') {
-            user.flair.ptrades.flair_css_class = css_class;
-          } else {
-            user.flair.svex.flair_css_class = css_class;
-          }
-          user.save(function (err) {
-            if (err) {
-              console.log(err);
-            }
-          });
-          Application.destroy({id: req.allParams().id}).exec(function (err, app) {
-            if (err) {
-              return res.serverError(err);
-            }
-            return res.ok(app);
-          });
-        }, function (err) {
-          return res.serverError(err);
-        });
-      });
-    });
+      var user = await User.findOne(app.user),
+        shortened = app.sub === 'pokemontrades' ? 'ptrades' : 'svex';
+      var relevant_flair = Flairs.makeNewCSSClass(_.get(user, 'flair.' + shortened + '.flair_css_class') || '', app.flair, app.sub);
+      user.flair[shortened].flair_css_class = relevant_flair;
+      await Reddit.setUserFlair(req.user.redToken, user.name, relevant_flair, user.flair[shortened].flair_text, app.sub);
+      var promises = [];
+      promises.push(user.save());
+      promises.push(Event.create({type: "flairTextChange", user: req.user.name,content: "Changed " + user.name + "'s flair to " + relevant_flair}));
+      var pmContent = 'Your application for ' + Flairs.formattedName(app.flair) + ' flair on /r/' + app.sub + ' has been approved.';
+      promises.push(Reddit.sendPrivateMessage(refreshToken, 'FlairHQ Notification', pmContent, user.name));
+      promises.push(Application.destroy({id: req.allParams().id}));
+      await* promises;
+      return res.ok(await Flairs.getApps());
+    } catch (err) {
+      return res.serverError(err);
```

Comment:
Maybe can we have multiple errors. 1 for if the flair isn't set correctly (and send that so we get the error and we can show it) and 1 for everything else (that we can sanitise if we want).